### ISSUE TITLE
Butterfly analogWrite and pinMode conflicts

### ISSUE THREAD

**Issue by jacky4566** [GrumpyOldPizza/arduino-STM32L4#48]
There seems to be a problem with calling pinMode multiple times and using analogWrite.
The following code will fail for me:

```c++
static const int MOSFET = 9;

void setup() {
  //ON
  pinMode(MOSFET, OUTPUT);
  analogWrite(MOSFET, 255);
  delay(1000);
  //OFF
  analogWrite(MOSFET, 0);
  delay(3000);
  //ON
  pinMode(MOSFET, OUTPUT);//This pinMode causes problems
  analogWrite(MOSFET, 255);
  delay(1000);
  //OFF
  analogWrite(MOSFET, 0);
  pinMode(MOSFET, INPUT);
}

void loop() {
  delay(1);
}
```

Quite a few of the Arduino universe libraries use this format so its a bit frustrating to change them all.

**Comment by GrumpyOldPizza** [comment]
What is the problem exactly ? Hang ? Wrong value ? Spike ?

**Comment by jacky4566** [comment]
Ah sorry. The code continues but no output voltage is observed.
The first "ON" executes correctly 3.3v for 1s but the second does nothing. 0 volts.

**Comment by kriswiner** [comment]
Why are you calling this a second time? Not necessary.

**Comment by jacky4566** [comment]
Yes I understand its not necessary but many Arduino libraries do this. This behavior is also very unexpected and could easily confuse new users. 
I found this while using an older [Motor Driver Library](https://www.arduinolibraries.info/libraries/motor-driver-library)

Its an easy fix to remove the pinMode() calls but still not great practice since pinMode() should be repeatable under Arduino. 

This also kills Charlieplexing LEDs where you need tri-state logic to change between INPUT and OUTPUT rapidly.

**Comment by kriswiner** [comment]
Changing the mode should work fine, repeatedly calling the same mode is no
good practice.

**Comment by jacky4566** [comment]
Changing the mode also does not work, see my example below. Repeatedly calling the same mode may not be good practice no, but it happens. [Even in Adafruit's libraries](https://github.com/adafruit/Adafruit-Motor-Shield-library)

I feel this should be a concern for you guys since it's common enough in the Arduino enviroment.

```c++
static const int MOSFET = 9;

void setup() {
  //ON
  pinMode(MOSFET, OUTPUT);
  analogWrite(MOSFET, 255);
  delay(1000);
  //OFF
  analogWrite(MOSFET, 0);
  delay(3000);
  //ON
  pinMode(MOSFET, INPUT);
  pinMode(MOSFET, OUTPUT);
  analogWrite(MOSFET, 255); //No output produced. 0v
  delay(1000);
  //OFF
  analogWrite(MOSFET, 0);
  pinMode(MOSFET, INPUT);
}

void loop() {
  delay(1);
}
```

**Comment by kriswiner** [comment]
Comment out these two lines and it should work:

pinMode(MOSFET, INPUT);
  pinMode(MOSFET, OUTPUT);

Why are you toggling the mode for no reason here?

If you insist on doing so, at least put a delay between the calls.

**Comment by jacky4566** [comment]
Kris the code is just a simplified example of how to cause the issue.
A delay of 50ms also did not fix the problem. 

Your free to do as you wish but I believe that repeated calls to "pinMode(  , OUTPUT);" should not cause a failure of the pin to conduct.
